src/BodyDigest.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace HttpSignatures;
+
+class BodyDigest
+{
+    /** @var string */
+    private $hashName;
+
+    /** @var string */
+    private $digestHeaderPrefix;
+
+    /**
+     * @param string $name
+     * @return BodyDigest
+     * @throws DigestException
+     */
+    public function __construct($hashAlgorithm = null)
+    {
+        switch (strtolower(str_replace("-", '', $hashAlgorithm))) {
+        case 'sha':
+        case 'sha1':
+            $this->hashName = 'sha1';
+            $this->digestHeaderPrefix = 'SHA';
+            break;
+        case 'sha256':
+        case null:
+        case '':
+            $this->hashName = 'sha256';
+            $this->digestHeaderPrefix = 'SHA-256';
+            break;
+        case 'sha512':
+            $this->hashName = 'sha512';
+            $this->digestHeaderPrefix = 'SHA-512';
+            break;
+        default:
+            throw new DigestException("Digest algorithm parameter '$hashAlgorithm' not understood");
+            break;
+        }
+    }
+
+    public function putDigestInHeaderList($headerList)
+    {
+        if (!array_search('digest', $headerList->names)) {
+            $headerList->names[] = 'digest';
+        };
+        return $headerList;
+    }
+
+    public function setDigestHeader($message)
+    {
+        $message = $message->withoutHeader('Digest')
+            ->withHeader(
+                'Digest',
+                $this->getDigestHeaderLinefromBody($message->getBody()));
+        return $message;
+    }
+
+    public function getDigestHeaderLinefromBody($messageBody)
+    {
+        if (is_null($messageBody)) {
+            $messageBody = '';
+        };
+        return $this->digestHeaderPrefix . '=' . base64_encode(hash($this->hashName, $messageBody, true));
+    }
+
+    public static function fromMessage($message)
+    {
+        if (! $digestLine = $message->getHeader('Digest')) {
+            throw new DigestException("No Digest header in message");
+        }
+        try {
+            return new BodyDigest(explode("=", $digestLine[0])[0]);
+        } catch (DigestException $e) {
+            throw $e;
+        }
```

Comment:
catching and immediately throwing makes the try-catch redundant

Reply:
Much more complete handling in 76c6bc0ecf5afbefa7e1ed2c49f8a4ba45919a20:
- An equals sign (`=`) is valid base64, so this explode may falsely work even on a badly formatted line
- Explicitly test if the content before the `=` is a valid hash spec
- Introduced a generic hash spec tester and list of valid hashes
- Cleaned up constructor to use the generic validator
- Throw an exception when logic says no, instead of a generic failure to parse
  - do or do not, there is no `try`